packages/client.api/src/queries/Queries.ts

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2024 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type {
+  ObjectQueryDataType,
+  ObjectSetQueryDataType,
+  OntologyDefinition,
+  QueryDataTypeDefinition,
+  QueryDefinition,
+} from "@osdk/api";
+import type {
+  BaseObjectSet,
+  DataValueClientToWire,
+  DataValueWireToClient,
+  NOOP,
+  ObjectSet,
+  OsdkBase,
+  OsdkObjectPrimaryKeyType,
+} from "../index.js";
+import type { PartialByNotStrict } from "../util/PartialBy.js";
+
+export type Queries<O extends OntologyDefinition<any>> = {
+  [K in keyof O["queries"]]: QuerySignatureFromDef<O["queries"][K]>;
+};
+
+export type QuerySignatureFromDef<T extends QueryDefinition<any, any>> =
+  keyof T["parameters"] extends never
+    ? () => Promise<QueryReturnType<T["output"]>>
+    : (
+      params: QueryParameterType<T["parameters"]>,
+    ) => Promise<QueryReturnType<T["output"]>>;
+
+export type QueryParameterType<
+  T extends Record<any, QueryDataTypeDefinition<any, any>>,
+> = NOOP<PartialByNotStrict<NotOptionalParams<T>, OptionalQueryParams<T>>>;
+
+export type QueryReturnType<T extends QueryDataTypeDefinition<any, any>> =
```

Comment:
I kept this separate from `BaseType` for now to keep things easier to parse through, but reading through this now I can probably just add another `R` to the generic to tell us if we have a return value

Reply:
Maybe something like:

```
```
type BaseType2<T extends QueryDataTypeDefinition<any, any>, R extends boolean> =
  T extends ObjectQueryDataType<any, infer TTargetType>
    ? R extends true ? OsdkBase<TTargetType>
    : OsdkBase<TTargetType> | OsdkObjectPrimaryKeyType<TTargetType>
    : T extends ObjectSetQueryDataType<any, infer TTargetType>
      ? R extends true ? ObjectSet<TTargetType> : BaseObjectSet<TTargetType>
    : T["type"] extends keyof DataValueClientToWire
      ? DataValueClientToWire[T["type"]]
    : never;
```
```